Autoproj separation

pierrewillenbrockdfki opened the pull request:
This separates out accesses to autoproj (and autobuild) internal structures. The patch series is split into preparatory work, mostly splitting functions in two, then copying functions over to the autoproj facing parts, implementing an information gathering function, moving over to using that function and finally cleaning up by removing the now unused functions.

pierrewillenbrockdfki committed:
deb: Replace Autoproj.{info|warn|error} with Autoproj::Packaging.{}

Better use our own logger.

pierrewillenbrockdfki committed:
deb: Add infrastructure for splitting out autoproj specific things

pierrewillenbrockdfki committed:
deb: Make root_dir settable

pierrewillenbrockdfki committed:
deb: Make TargetPlatform output as Autoproj::Packaging

I think it does not have access to the Autoproj::Packaging::Packagers
Logger implementation.

pierrewillenbrockdfki committed:
deb: Replace alias system, only alias commandline packages

Uses autoproj.moved_packages instead

pierrewillenbrockdfki committed:
deb: Stop looking for Autoproj.manifest in Autoproj::Packaging

pierrewillenbrockdfki committed:
deb: Change variable name to dependency_debian_name

pierrewillenbrockdfki committed:
deb: Split out filtered_dependencies

pierrewillenbrockdfki committed:
deb: Split filter_all_required_packages out of all_required_packages

This has the side-effect of not filtering the rock packages when determining
gem_versions.

pierrewillenbrockdfki committed:
deb: Remodel convert_gem to not require access to autoproj

pierrewillenbrockdfki committed:
deb: Handle parallel_build_level in build_local

pierrewillenbrockdfki committed:
deb: debian_plain_version does not require a distribution parameter

Its probably a good idea to cache local_commit_time, but thats going away
soon.

pierrewillenbrockdfki committed:
deb: Pregenerate some more information before expanding templates

pierrewillenbrockdfki committed:
deb: Change autoproj1adaptors logger

pierrewillenbrockdfki committed:
deb: filtered_dependencies is not available in autoproj1adaptor

pkgi.dependencies must be passed through filtered_dependencies at
a later point.

pierrewillenbrockdfki committed:
deb: Add a function to extract the required info about packages from autoproj

pierrewillenbrockdfki committed:
deb: copy things over to autoproj1adaptor

pierrewillenbrockdfki commented:
I failed to check through the meta related changes, there are some things that need to change there, too.

pierrewillenbrockdfki commented:
This version works with meta packages (as in: it does not crash and burn)

2maz commented on bin/deb_local:
```diff
@@ -67,7 +68,7 @@ def build_locally(packager, pkg_name, options)
         end
         puts "############### build #{pkg_name} #####################"
-        debian_package_dir = packager.packaging_dir(options[:debian_package_name])
+        debian_package_dir = packager.packaging_dir_i(options[:debian_package_name])
```
I would prefer if you cleanup the use of _i  and replacing it by 'info', e.g. for pkgi to pkginfo, where it makes sense.
In this case for example, I would remain with packaging_dir

pierrewillenbrockdfki replied:
packaging_dir also allows a PackageInfo description. The conversion _i to _info and pkgi to pkginfo is easily done. I was wondering if it is better to just drop the suffix? Would decrease the size of the diff substantially, but removes the information where different objects are passed around.

2maz replied:
as I said, where is make sense you can keep it,e.g. if pkg really means pkginfo. Otherwise remove it, like for packaging_dir.

2maz commented on bin/deb_local:
```diff
@@ -187,15 +188,19 @@ end
 end
 
 packager = Autoproj::Packaging::Debian.new(build_options)
+autoprojadaptor = Autoproj::Packaging::Autoproj1Adaptor.new(build_options)
```
To prepare for multiple adaptor implementations, e.g., autoproj does not need to be the only provider for pkg_info, add a parent (abstract) like class PkgInfoAsk
ask = Autoproj::Packaging::PkgInfoAsk.new(build_options, :autoproj_v1)

pierrewillenbrockdfki replied:
Do you want me to add this to this pull request, i.E. remove dependencies on autoproj from deb_local, deb_package, lib/rock/packaging/jenkins etc as well?

2maz replied:
if possible yes, remove the direct dependencies

2maz commented on bin/deb_local:
```diff
@@ -256,25 +261,28 @@ Autoproj.silent = true
 root_dir  = Autoproj::CmdLine.initialize_root_directory
```
Should be encapsulated in the Autoproj1Adaptor as well

2maz commented on bin/deb_local:
```diff
@@ -381,7 +389,8 @@ else
             end
             pkg_build_options = build_options.dup
-            pkg_build_options[:debian_package_name] = packager.debian_name(pkg,true)
+            pkgi = autoprojadaptor.pkgi_from_pkg(pkg)
```
pkgi -> pkginfo

2maz commented on bin/deb_package:
```diff
@@ -326,37 +331,44 @@ if !o_architectures.empty?
     options[:architecture] = o_architectures.first
 end
 packager = Autoproj::Packaging::Debian.new(options)
+autoprojadaptor = Autoproj::Packaging::Autoproj1Adaptor.new(options)
```
see comments on deb_local

2maz commented on lib/rock/packaging/templates/debian/changelog:
```diff
@@ -1,23 +1,8 @@
-    * the repository and commit information could not be extracted
-      * error at generation: <%= e.to_s %>
-<%end%>
+<% origin_information.each do |line| %>
```
some more information/documentation on origin information is needed

pierrewillenbrockdfki replied:
Documented in packageinfo.rb

2maz commented on lib/rock/packaging/templates/debian/control:
```diff
@@ -2,32 +2,6 @@ Source: <%= debian_name %>
 Section: science
 Priority: extra
 Maintainer: Rock Packaging Daemon <rock-dev@dfki.de>
-% dependencies = (deps_rock_packages + deps_osdeps_packages + deps_nonnative_packages).flatten
```
where did adding these dependencies go? or is this obsolete?

pierrewillenbrockdfki replied:
That was moved to generate_debian_dir. dependencies is only used to when generating build_dependencies.

pierrewillenbrockdfki committed:
deb: Remove _i suffix of functions duplicated earlier

pierrewillenbrockdfki commented:
Stopped using partial hungarian notation for the function names(dropped the _i indicating PackageInfo version), but instead uses pkginfo at all places applicable.
Moved more functionality into the autoproj adaptor.
In addition, i put the fixes for meta package generation on top.

pierrewillenbrockdfki committed:
deb: Change into the gems directory when parsing the spec file

Spec files can contain ruby snippets that expect to execute in
the gems directory.

2maz commented on bin/deb_local:
```diff
@@ -191,11 +192,15 @@ build_options[:release_name] ||= 'local'
     exit 0
 end
 
+autoprojadaptor = Autoproj::Packaging::PackageInfoAsk.new(:detect, Hash.new())
```
autoprojadaptor --> package_info_ask

2maz commented on bin/deb_local:
```diff
@@ -252,59 +257,61 @@ if o_deregister
-Autoproj.silent = true
-root_dir  = Autoproj::CmdLine.initialize_root_directory
-selection = Autoproj::CmdLine.initialize_and_load(nil)
+autoprojadaptor.init_autoproj
```
that should either be part of the initialize in the pkg_info_ask or rename function to init_database

2maz commented on bin/deb_local:
```diff
@@ -252,59 +257,61 @@ if o_deregister
 
 packager.rock_release_name = build_options[:release_name]
 build_options[:packager] = packager
+build_options[:autoprojadaptor] = autoprojadaptor
```
autoprojadaptor --> rename

2maz commented on bin/deb_package:
```diff
@@ -325,37 +331,42 @@ end
-Autoproj.silent = true
-root_dir  = Autoproj::CmdLine.initialize_root_directory
-selection = Autoproj::CmdLine.initialize_and_load(nil)
+autoprojadaptor.init_autoproj
```
s. comments above

2maz commented on lib/rock/packaging/packageinfoask.rb:
```diff
@@ -0,0 +1,177 @@
+            end
+
+            # returns true if pkgname is an autoproj meta package
+            def ismetapackage?(package_name)
```
prefer: is_metapackage?

2maz commented on lib/rock/packaging/packageinfo.rb:
```diff
@@ -0,0 +1,62 @@
+            attr_accessor :origin_information
+            # Number of parallel build processes
+            attr_accessor :parallel_build_level
+            # importer type, one of :git, :svn, :archiveimporter
```
prefer: :archive_importer
generally: from CamelCase to camel_case

2maz commented on lib/rock/packaging/packageinfoask.rb:
```diff
@@ -0,0 +1,177 @@
+            class << self
+                alias :class_new :new
+
+                def new(which, options)
```
which not using initialize here?

pierrewillenbrockdfki replied:
Cannot return a object of a different class from initialize, just initialize it.

2maz replied:
Ok, I should look closer ;). Why did you opt for this solution vs. using initialize with an optional parameter for the selection of the particular adaptor implementation, and delegating the query calls?

pierrewillenbrockdfki replied:
PackageInfoAsk is an abstract base class to the package info providers, and thus mostly there for documentation purposes. The new() method has no real purpose in abstract classes, and, together with the initialize method, would generally raise an exception, if not for the factory semantics new() has now. If that new() override is problematic, i'd rather have a differently named class method do the instantiation and have initialize raise an exception than doing basically \@delegate.method() for every method, or going the method_missing route(which would require to remove all method defs).

pierrewillenbrockdfki committed:
deb: tests: looks like utilrb does in fact not depend on "rake-compiler", "ruby-flexmock"

flexmock comes in as a "test_depend", no idea what it is about rake-compiler

pierrewillenbrockdfki committed:
deb: Move @ruby_gems, @ruby_rock_gems, @osdeps out of Debian packager

These hidden globals should not be hidden and make dependency calculation
have unintended side effects.

pierrewillenbrockdfki committed:
deb: Use autoproj1adaptor instead of packager where the former has functionality

pierrewillenbrockdfki committed:
deb: Move patching to the initial tar file generation in convert_gem

pierrewillenbrockdfki committed:
deb: tests: 'mustermannwebsocket-driver' => 'mustermann','websocket-driver'

pierrewillenbrockdfki reviewed (commented):
Commit "deb: Move @ruby_gems, @ruby_rock_gems, @osdeps out of Debian packager"
probably should be reviewed with some more care, it changes some semantics, and keeps some questionable semantics.

pierrewillenbrockdfki committed:
deb: Fix patch_dir in deb_local

Should be o_patch_dir

pierrewillenbrockdfki committed:
deb: Fix debian/rules template in the presence of bundles

pierrewillenbrockdfki committed:
deb: Fix debian/rules template in the presence of bundles

pierrewillenbrockdfki committed:
deb: Improve mangling of include paths in presence of quotes

pierrewillenbrockdfki committed:
deb: tests: Handle debian.tar.xz in addition to debian.tar.gz

pierrewillenbrockdfki committed:
deb: Also patch the initial tar file generation in convert_gem

dh-make-ruby chokes on malformed spec files

pierrewillenbrockdfki committed:
deb: Fix meta template

There is no changelog information for that thing.

pierrewillenbrockdfki committed:
deb: make postinst create the doc dir as well

otherwise fails when the package does not actually install docs and there
is no other package that installed docs, yet.

pierrewillenbrockdfki committed:
deb: ruby gem version can also have .rc##

pierrewillenbrockdfki committed:
deb: Fix TEMPLATES_DIR

pierrewillenbrockdfki committed:
deb: Import packages before trying to find commit times or load manifest

pierrewillenbrockdfki committed:
deb: Add cache for manifests

pierrewillenbrockdfki committed:
deb: Fix svn commit time fetch

pierrewillenbrockdfki committed:
deb: Remove unused functions and variables

pierrewillenbrockdfki committed:
deb: remove leftover rock_release_platform generation

every time rock_release_name is updated, rock_release_platform is updated
as well.

pierrewillenbrockdfki committed:
deb: Fixes for meta package building

pierrewillenbrockdfki committed:
deb: Move most of the other references to autoproj into the adaptor

Notable exceptions are Autoproj.osdeps.resolve_package,
Autoproj::PackageManagers::AptDpkgManager in jenkins.rb,
::Autoproj::PackageManagers::GemManager.new in gem_dependencies.rb
Autoproj.manifest.metapackages in deb_local, but that is going away soon
